### ISSUE TITLE
Create Bayes by Backprop Op

### ISSUE THREAD

**Issue by aidan-plenert-macdonald** [aidan-plenert-macdonald/incubator-mxnet#1]
Using [Dense](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/nn/basic_layers.py#L177) as an example, we can create a Hybrid op for BBB training,

```
class NoiseDense(gluon.nn.Dense):
    def __init__(self, *args, **kwargs):
        super(NoiseDense, self).__init__(*args, **kwargs)
        self.weight_std = self.params.get('weight_std', shape=(self._units, self._in_units),
                                          init=kwargs['weight_initializer'],
                                          allow_deferred_init=True)
        if use_bias:
            self.bias_std = self.params.get('bias_std', shape=(self._units,),
                                            init=kwargs['bias_initializer'],
                                            allow_deferred_init=True)
    
    def hybrid_forward(self, F, x, weight, bias=None):
        act = F.FullyConnected(x, weight, bias, no_bias=bias is None, num_hidden=self._units,
                               flatten=self._flatten, name='fwd')
        if self.act is not None:
            act = self.act(act)
        return act
```

Unfortunately, the [`hybrid_forward`](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/nn/basic_layers.py#L200) functions use [built in C++](https://github.com/apache/incubator-mxnet/blob/master/src/operator/fully_connected-inl.h#L74) to compute the forward and backprop in the hybrid case. I believe I may need to define one in C++ because the hybrid forward function takes as input the weights already.

Although the [call to to `hybrid_forward`](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/block.py#L432) doesn't appear to pass the weight explicitly and furthermore, [Dropout's `hybrid_forward`](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/nn/basic_layers.py#L273) doesn't look for weights. So I may be in luck and just need to receive both `weights` and `weights_std` to be able to create by block.

**Comment by aidan-plenert-macdonald** [comment]
Mostly functioning code. https://github.com/aidan-plenert-macdonald/mxnet-the-straight-dope/blob/bayes-backprop/chapter18_variational-methods-and-uncertainty/bayes-by-backprop-gluon.ipynb
